### DISCUSSION TITLE
aiortc with OpenCV imshow

### DISCUSSION THREAD

**Discussion by ahmadmughees** [aiortc/aiortc#734] [Q&A]
Hi, Thanks for amazing framework, 

I just encounter a bug when using aiortc and OpenCV imshow fuction. 

Whenever I even simply import the aiortc, my program hangs on the opencv imshow function. Minimum example is as follows, 

```import cv2
import numpy as np
import aiortc # if you comment this line, the program will work fine

img = 255 * np.ones((512, 1920 // 2, 3), dtype=np.uint8)
for i in range(1000):
    cv2.imshow("win", img)
    print(i)
    cv2.waitKey(1)
```
The library versions are
```
aiortc==1.3.2
opencv-python==4.5.5.64
```
Thanks alot.

**Comment by abhiTronix** [comment]
@ahmadmughees I can't produce this error on any of my machines. I get this output with your code:
```sh
➜ py k.py
0
1
2
3
4
5
6
7
8
9
10
11
12
13
14
15
16
17
18
19
20
21
22
23
24
25
26
27
28
29
30

------

997
998
999
```

And here're the exact versions:

```sh
➜ pip show opencv-python
Name: opencv-python
Version: 4.5.5.64
Summary: Wrapper package for OpenCV python bindings.
Home-page: https://github.com/skvark/opencv-python
Author:
Author-email:
License: MIT
Location: c:\users\abhiu\appdata\roaming\python\python38\site-packages
Requires: numpy, numpy
Required-by:

➜ pip show aiortc
Name: aiortc
Version: 1.3.2
Summary: An implementation of WebRTC and ORTC
Home-page: https://github.com/aiortc/aiortc
Author: Jeremy Lainé
Author-email: jeremy.laine@m4x.org
License: BSD
Location: c:\users\abhiu\appdata\roaming\python\python38\site-packages
Requires: aioice, av, cffi, cryptography, google-crc32c, pyee, pylibsrtp
Required-by:
```



Something else is wrong with your setup.

**Comment by Compiler** [comment]
I also have the same problem. It doesn't have to be aiortc directly, any reference to the library I have seen won't allow imshow to work for me

**Comment by maiermic** [comment]
@ahmadmughees @abhiTronix @Compiler Which operating system are you using?

PS: Just an arbitrary idea: Try replacing `opencv-python` with version `4.5.4.60` or `opencv-contrib-python`.

**Reply by Compiler** [reply]
Thanks for the idea. Some new info after debugging is that it works on windows w/ any versions and I couldn't get most older opencv versions to build

opencv-contrib-python didn't work and opencv-python==4.5.4.60 also didn't work.

They both do the same thing of hanging on imshow and never showing a window and then crashing the terminal. 

But this is only linux, windows works w/ all combinations of versions that i tested.

**Reply by maiermic** [reply]
You could try importing only individual submodules from aiortc. Thereby, you should be able to narrow down the cause.

**Comment by maiermic** [comment]
@ahmadmughees Your example is missing an import of `cv2`. I'm wondering, why nobody mentioned this yet. Did everyone fix this silently? Have you tried chaning the order of the imports?

**Comment by maiermic** [comment]
It runs (on Linux) if the window is created before the import of `aiortc`, e.g.

```python
import cv2
cv2.namedWindow("win", cv2.WINDOW_NORMAL)

import numpy as np
import aiortc # if you comment this line, the program will work fine

img = 255 * np.ones((512, 1920 // 2, 3), dtype=np.uint8)
for i in range(1000):
    cv2.imshow("win", img)
    print(i)
    cv2.waitKey(1)
```

You can replace `imshow` with a call to `namedWindow` and it will block the execution. too

```python
import cv2
import numpy as np
import aiortc # if you comment this line, the program will work fine

img = 255 * np.ones((512, 1920 // 2, 3), dtype=np.uint8)
for i in range(1000):
    cv2.namedWindow("win", cv2.WINDOW_NORMAL)  # blocks execution
    print(i)
    cv2.waitKey(1)
```

This is not caused by the lock on the window mutex (see source of [imshow](https://github.com/opencv/opencv/blob/2104d61d4a3fb363e5c7d29a8d14eac038e2b3b2/modules/highgui/src/window.cpp#L97) and [namedWindow](https://github.com/opencv/opencv/blob/2104d61d4a3fb363e5c7d29a8d14eac038e2b3b2/modules/highgui/src/window.cpp#L452)). Otherwise, [destroyAllWindows](https://github.com/opencv/opencv/blob/2104d61d4a3fb363e5c7d29a8d14eac038e2b3b2/modules/highgui/src/window.cpp#L508) should also block the execution, but it doesn't.

Nevertheless, the source might give a hint on the cause of the issue :nerd_face:

---

Some other things I noticed:

1. Calling
    ```python
    cv2.namedWindow('win', -1)
    ```
    after an import of `aiortc` does block. Without the import, an error is thrown
    ```
    Traceback (most recent call last):
      File "example.py", line 6, in <module>
        cv2.namedWindow('win', -1)
    cv2.error: OpenCV(4.6.0) /io/opencv/modules/highgui/src/window_QT.cpp:1705: error: (-218:No OpenGL support) Library was built without OpenGL support in function 'CvWindow'
    ```
    That means that the cause of this error message is not reached, when the import of `aiortc` is present.
2. If I set the environment variables
    ```bash
    export OPENCV_LOG_LEVEL=DEBUG
    export OPENCV_VIDEOIO_DEBUG=1
    ```
    and run
    ```python
    import cv2
    import numpy as np
    import aiortc # if you comment this line, the program will work fine
    
    img = 255 * np.ones((512, 1920 // 2, 3), dtype=np.uint8)
    cv2.namedWindow("win", cv2.WINDOW_NORMAL)  # blocks execution
    ```
    I get the log
    ```
    [DEBUG:0@0.856] global /io/opencv/modules/highgui/src/backend.cpp (120) createDefaultUIBackend UI: Initializing backend...
    [DEBUG:0@0.856] global /io/opencv/modules/highgui/src/registry.impl.hpp (87) UIBackendRegistry UI: Builtin backends(3): GTK(1000); GTK3(990); GTK2(980) + BUILTIN(QT5)
    [DEBUG:0@0.856] global /io/opencv/modules/highgui/src/registry.impl.hpp (112) UIBackendRegistry UI: Available backends(3): GTK(1000); GTK3(990); GTK2(980) + BUILTIN(QT5)
    [ INFO:0@0.856] global /io/opencv/modules/highgui/src/registry.impl.hpp (114) UIBackendRegistry UI: Enabled backends(3, sorted by priority): GTK(1000); GTK3(990); GTK2(980) + BUILTIN(QT5)
    [DEBUG:0@0.856] global /io/opencv/modules/highgui/src/backend.cpp (78) createUIBackend UI: trying backend: GTK (priority=1000)
    [DEBUG:0@0.856] global /io/opencv/modules/highgui/src/plugin_wrapper.impl.hpp (220) getPluginCandidates UI: GTK plugin's glob is 'libopencv_highgui_gtk*.so', 1 location(s)
    [DEBUG:0@0.856] global /io/opencv/modules/highgui/src/plugin_wrapper.impl.hpp (230) getPluginCandidates     - /path/to/site-packages/cv2: 0
    [DEBUG:0@0.856] global /io/opencv/modules/highgui/src/plugin_wrapper.impl.hpp (234) getPluginCandidates Found 0 plugin(s) for GTK
    [DEBUG:0@0.856] global /io/opencv/modules/highgui/src/backend.cpp (78) createUIBackend UI: trying backend: GTK3 (priority=990)
    [DEBUG:0@0.856] global /io/opencv/modules/highgui/src/plugin_wrapper.impl.hpp (220) getPluginCandidates UI: GTK3 plugin's glob is 'libopencv_highgui_gtk3*.so', 1 location(s)
    [DEBUG:0@0.856] global /io/opencv/modules/highgui/src/plugin_wrapper.impl.hpp (230) getPluginCandidates     - /path/to/site-packages/cv2: 0
    [DEBUG:0@0.856] global /io/opencv/modules/highgui/src/plugin_wrapper.impl.hpp (234) getPluginCandidates Found 0 plugin(s) for GTK3
    [DEBUG:0@0.856] global /io/opencv/modules/highgui/src/backend.cpp (78) createUIBackend UI: trying backend: GTK2 (priority=980)
    [DEBUG:0@0.856] global /io/opencv/modules/highgui/src/plugin_wrapper.impl.hpp (220) getPluginCandidates UI: GTK2 plugin's glob is 'libopencv_highgui_gtk2*.so', 1 location(s)
    [DEBUG:0@0.856] global /io/opencv/modules/highgui/src/plugin_wrapper.impl.hpp (230) getPluginCandidates     - /path/to/site-packages/cv2: 0
    [DEBUG:0@0.856] global /io/opencv/modules/highgui/src/plugin_wrapper.impl.hpp (234) getPluginCandidates Found 0 plugin(s) for GTK2
    [DEBUG:0@0.856] global /io/opencv/modules/highgui/src/backend.cpp (106) createUIBackend UI: fallback on builtin code: QT5
    ```
    I get the same log if I remove the import of `aiortc`

**Comment by maiermic** [comment]
The issue is caused by an import of `av` in `__init__.py` of `aiortc`. You get the same issue without `aiortc`

```python
import cv2
import numpy as np
import av # if you comment this line, the program will work fine

img = 255 * np.ones((512, 1920 // 2, 3), dtype=np.uint8)
cv2.namedWindow("win", cv2.WINDOW_NORMAL)  # blocks execution
```

PyAV's logging is disabled, but still seems to cause a deadlock, see https://github.com/aiortc/aiortc/blob/main/src/aiortc/__init__.py#L52-L53

The deadlock remains, even if

```python
export PYAV_LOGGING=off
```

is set.

---

| Package               | Version |
| --------------------- | -------- |
| av                    | 10.0.0 |
| numpy                 | 1.24.3 |
| opencv-contrib-python | 4.7.0.72 |

**Comment by ashishmiryalkar** [comment]
any updates on this topic?

**Comment by tomotake** [comment]
before import aiortc
you can do

img = np.zeros((h,w,3), dtype=np.uint8)
cv2.imshow('video', img)
cv2.waitKey(1)

then  import aiortc
use the same name 'video' and use cv2.imshow(), cv2.waitKey(1), this time, it works.

**Comment by tomotake** [comment]
I found this is the issue of pip installed av package. Please try one of the following. It sovled my issue.
pip install av --no-binary av
or
conda install -c conda-forge av
reference ... https://stackoverflow.com/questions/72604912/cant-show-image-with-opencv-when-importing-av
